alibi/prototypes/protoselect.py

```diff
@@ -0,0 +1,543 @@
+            'n_splits': n_splits,
+            'batch_size': batch_size,
+            'scores': scores,
+        }
+    }
+
+
+def _batch_preprocessing(X: np.ndarray,
+                         preprocess_fn: Callable[[np.ndarray], np.ndarray],
+                         batch_size: int = 32) -> np.ndarray:
+    """
+    Preprocess a dataset `X` in batches by applying the preprocessor function.
+
+    Parameters
+    ----------
+    X
+        Dataset to be preprocessed.
+    preprocess_fn
+        Preprocessor function.
+    batch_size
+        Batch size to be used for each call to `preprocess_fn`.
+
+    Returns
+    -------
+    Preprocessed dataset.
+    """
+    X_ft = []
+    num_iter = int(np.ceil(len(X) / batch_size))
+
+    for i in range(num_iter):
+        istart, iend = batch_size * i, min(batch_size * (i + 1), len(X))
+        X_ft.append(preprocess_fn(X[istart:iend]))
+
+    return np.concatenate(X_ft, axis=0)
+
+
+def _imscatterplot(x: np.ndarray,
+                   y: np.ndarray,
+                   images: np.ndarray,
+                   figsize: Tuple[int, int],
+                   image_size: Tuple[int, int] = (28, 28),
+                   zoom: Optional[np.ndarray] = None,
+                   zoom_lb: float = 1.0,
+                   zoom_ub=2.0,
+                   sort_by_zoom: bool = True) -> None:
+    """
+    2D image scatter plot.
+
+    Parameters
+    ----------
+    x
+        Images' x-coordinates.
+    y
+        Images' y-coordinates.
+    images
+        Array of images to be placed at coordinates `(x, y)`.
+    figsize
+        `Matplotlib` figure size.
+    image_size
+        Size of the generated output image as `(rows, cols)`.
+    zoom
+        Images' zoom to be used.
+    zoom_lb
+        Zoom lower bound. The zoom values will be scaled linearly between `[zoom_lb, zoom_up]`.
+    zoom_ub
+        Zoom upper bound. The zoom values will be scaled linearly between `[zoom_lb, zoom_up]`.
+    """
+    if zoom is None:
+        zoom = np.ones(len(images))
+
+    zoom_min, zoom_max = np.min(zoom), np.max(zoom)
+    zoom = (zoom - zoom_min) / (zoom_max - zoom_min) * (zoom_ub - zoom_lb) + zoom_lb
+
+    if sort_by_zoom:
+        idx = np.argsort(zoom)[::-1]  # type: ignore
+        zoom = zoom[idx]  # type: ignore
+        x, y, images = x[idx], y[idx], images[idx]
+
+    fig, ax = plt.subplots(figsize=figsize)
+    ax.set_xticks([])
+    ax.set_yticks([])
+
+    resized_imgs = [resize(images[i], image_size) for i in range(len(images))]
+    imgs = [OffsetImage(img, zoom=zoom[i], cmap='gray') for i, img in enumerate(resized_imgs)]  # type: ignore
+    artists = []
+
+    for i in range(len(imgs)):
+        x0, y0, im = x[i], y[i], imgs[i]
+        ab = AnnotationBbox(im, (x0, y0), xycoords='data', frameon=False)
+        artists.append(ax.add_artist(ab))
+
+    ax.update_datalim(np.column_stack([x, y]))
+    ax.autoscale()
+
+
+def visualize_prototypes(explanation: 'Explanation',
```

Comment:
If relevant, consider if we can take some implementation details from `plot_ale`: https://github.com/SeldonIO/alibi/blob/61f0c6afd7e5e23585fc3661b4a4aa137db9abb7/alibi/explainers/ale.py#L455

E.g. we might want to return the axes for further customization by the user if desired?

Reply:
Or be more consistent with arguments affecting the figure/axes, e.g. for `plot_ale` I exposed `fig_kw` dictionary which allows users greater control. Passing a figsize then could be done as inhttps://docs.seldon.io/projects/alibi/en/latest/examples/ale_regression_boston.html#ALE-for-the-linear-regression-model

Reply:
Made some changes.